pkg/controller/networkpolicy/clusternetworkpolicy.go

```diff
@@ -0,0 +1,267 @@
+	// Old and current NetworkPolicy share the same key.
+	key, _ := keyFunc(oldCNP)
+	// Lock access to internal NetworkPolicy store such that concurrent access
+	// to an internal NetworkPolicy is not allowed. This will avoid the
+	// case in which an Update to an internal NetworkPolicy object may
+	// cause the SpanMeta member to be overridden with stale SpanMeta members
+	// from an older internal NetworkPolicy.
+	n.internalNetworkPolicyMutex.Lock()
+	oldInternalNPObj, _, _ := n.internalNetworkPolicyStore.Get(key)
+	oldInternalNP := oldInternalNPObj.(*antreatypes.NetworkPolicy)
+	// Must preserve old internal NetworkPolicy Span.
+	curInternalNP.SpanMeta = oldInternalNP.SpanMeta
+	n.internalNetworkPolicyStore.Update(curInternalNP)
+	// Unlock the internal NetworkPolicy store.
+	n.internalNetworkPolicyMutex.Unlock()
+	// Enqueue addressGroup keys to update their Node span.
+	for _, rule := range curInternalNP.Rules {
+		for _, addrGroupName := range rule.From.AddressGroups {
+			n.enqueueAddressGroup(addrGroupName)
+		}
+		for _, addrGroupName := range rule.To.AddressGroups {
+			n.enqueueAddressGroup(addrGroupName)
+		}
+	}
+	n.enqueueInternalNetworkPolicy(key)
+	for _, atg := range oldInternalNP.AppliedToGroups {
+		// Delete the old AppliedToGroup object if it is not referenced
+		// by any internal NetworkPolicy.
+		n.deleteDereferencedAppliedToGroup(atg)
+	}
+	n.deleteDereferencedAddressGroups(oldInternalNP)
+}
+
+// deleteCNP receives ClusterNetworkPolicy DELETED events and deletes resources
+// which can be consumed by agents to delete corresponding rules on the Nodes.
+func (n *NetworkPolicyController) deleteCNP(old interface{}) {
+	cnp, ok := old.(*secv1beta1.ClusterNetworkPolicy)
+	if !ok {
+		tombstone, ok := old.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.Errorf("Error decoding object when deleting ClusterNetworkPolicy, invalid type: %v", old)
+			return
+		}
+		cnp, ok = tombstone.Obj.(*secv1beta1.ClusterNetworkPolicy)
+		if !ok {
+			klog.Errorf("Error decoding object tombstone when deleting ClusterNetworkPolicy, invalid type: %v", tombstone.Obj)
+			return
+		}
+	}
+	defer n.heartbeat("deleteCNP")
+	klog.Infof("Processing ClusterNetworkPolicy %s DELETE event", cnp.Name)
+	key, _ := keyFunc(cnp)
+	oldInternalNPObj, _, _ := n.internalNetworkPolicyStore.Get(key)
+	oldInternalNP := oldInternalNPObj.(*antreatypes.NetworkPolicy)
+	klog.Infof("Old internal NetworkPolicy %#v", oldInternalNP)
+	err := n.internalNetworkPolicyStore.Delete(key)
+	if err != nil {
+		klog.Errorf("Error deleting internal NetworkPolicy during NetworkPolicy %s delete: %v", cnp.Name, err)
+		return
+	}
+	for _, atg := range oldInternalNP.AppliedToGroups {
+		n.deleteDereferencedAppliedToGroup(atg)
+	}
+	n.deleteDereferencedAddressGroups(oldInternalNP)
+}
+
+// toAntreaServicesForCrd converts a secv1beta1.NetworkPolicyPort object to an
+// Antrea Service object.
+func toAntreaServicesForCrd(npPorts []secv1beta1.NetworkPolicyPort) []networking.Service {
```

Comment:
nit: toAntreaServicesForCRD to keep align with existing style?

Reply:
done